feat(android): Expose api to handle Android 12 REQUIRE_EXACT_ALARM permission

vincent-paing opened the pull request:
Android 12 introduces new permission that is required before scheduling alarms. See: https://developer.android.com/about/versions/12/behavior-changes-12#exact-alarm-permission

This PR adds the followings:
- Expose `canScheduleExactAlarms` in the library with default back to `true` on iOS and Android < 12. 
- Expose a function to allow user to open Alarms & Reminders Settings. This defaults back to auto resolve on iOS and Android < 12.
- Added a broadcast receiver to reschedule triggers when user allows the permission.
- Updated trigger doc with the information on Android 12 limitations.

Concerns 
- Currently `AlarmPermissionBroadcastReceiver` has same functionality as `RebootBroadcastReceiver`. Should we just add one more intent filter to `RebootBroadcastReceiver` instead?
- Should the exact alarm permission be included in `getNotificationSettings`. We can add a new `AndroidSettings` type to the return type of that function, and include this boolean flag in that object.

Addresses #239

mikehardy commented on packages/react-native/src/types/Module.ts:
```diff
@@ -184,6 +184,31 @@ export interface Module {
   displayNotification(notification: Notification): Promise<string>;
 
+  /**
+   * API used to check whether you can create trigger notification with exact timestamp on Android >= 12.
```
I noticed that you use Android version in one spot and Android API number in another, perhaps best to specify both in both locations? Just in case - react-native devs don't always know everything about android versioning

```suggestion
   * API used to check whether you can create trigger notification with exact timestamp on Android >= 12 / API >= 31
```

mikehardy commented on packages/react-native/src/types/Module.ts:
```diff
@@ -184,6 +184,31 @@ export interface Module {
+   *
+   * on iOS, this will default to true
+   *
+   * If the API version is < 31, this will default to true
```
```suggestion
   * For Android < 12 / API < 31, this will default to true
```

mikehardy commented on packages/react-native/src/types/Module.ts:
```diff
@@ -184,6 +184,31 @@ export interface Module {
+  /**
+   * API used to open the Android Alarm special access settings for the application.
+   *
+   * If the API version is >= 31, the alarm special acess settings screen is displayed, otherwise,
```
```suggestion
   * On Android >= 12 / API >= 31, the alarm special access settings screen is displayed, otherwise,
```

mikehardy commented on docs/react-native/docs/triggers.md:
```diff
@@ -143,16 +143,32 @@ const trigger: TimestampTrigger = {
+
+### Android 12 Limitations
+
+Starting from Android 12, timestamp triggers cannot be created unless user specfically allow the [exact alarm permission](https://developer.android.com/about/versions/12/behavior-changes-12).
```
I don't feel really strongly here, but I slightly prefer this link - it contains a link to more info which is the original link you had here, but this is the stable documentation reference in case the behavior changes docs go away or move for whatever reason?

```suggestion
Starting from Android 12, timestamp triggers cannot be created unless user specfically allow the [exact alarm permission](https://developer.android.com/reference/android/Manifest.permission#SCHEDULE_EXACT_ALARM).
```

vincent-paing replied:
That's great and the new link also has reference to behavior changes in itself, so it should suffice

mikehardy commented on docs/react-native/docs/triggers.md:
````diff
@@ -143,16 +143,32 @@ const trigger: TimestampTrigger = {
+}
+```
+
+Please note that when the user revoke the permission, all of the timestamp triggers will be deleted by the system. However, notifee will automatically reschedules these triggers when the user allows the alarm permission again with no need for additional code.
````
```suggestion
Please note that if the user revokes the permission via system preferences, all of the timestamp triggers will be deleted by the system. However, if you check for the permission, notice it is missing, educate the user and they grant permission again, notifee will automatically reschedule the triggers when the user allows the alarm permission again with no need for additional code.
```

mikehardy commented on docs/react-native/docs/triggers.md:
```diff
@@ -143,16 +143,32 @@ const trigger: TimestampTrigger = {
+if (Notifee.canCreateTimeStampTrigger()) {
+  //Create timestamp trigger
+} else {
+  await Notifee.openAlarmPermissionSettings();
```
```suggestion
  // Show some user information to educate them on what exact alarm permission is,
  // and why it is necessary for your app functionality, then send them to system preferences:
  await Notifee.openAlarmPermissionSettings();
```

mikehardy reviewed (changes requested):
Wow, this is fantastic! Thank you @vincent-paing 
I left some really minor comments, all documentation / wording related. Your code looks basically spotless :sparkles: 

What do you think on those? I'll switch to approve after resolving those and that should be easy I think

You had some questions:

>  Currently AlarmPermissionBroadcastReceiver has same functionality as RebootBroadcastReceiver. Should we just add one more intent filter to RebootBroadcastReceiver instead?

If we were going from one receiver to two, it might make more sense to make them generic. But given we already have 4 receivers if I could correctly, I don't think it is a big deal. I don't feel strongly about this, so if it's already tested and working, I am fine with it personally the way it is

> Should the exact alarm permission be included in getNotificationSettings. We can add a new AndroidSettings type to the return type of that function, and include this boolean flag in that object.

Interesting idea. I think it would be sort of "syntactic sugar" since it would be equivalent to just calling getNotificationSettings followed by canCreateTimestampTrigger, but since every app will likely need to call both of those all the time now, this idea seems like it would improve developer experience. I think the PR could go as is, but if you wanted to extend NotificationSettings to have a new android key and inside that have an AndroidNotificationSettings canCreateTimestampTrigger: boolean (or whatever seemed most correct!) that would be nice

I'm forever curious for whatever thoughts @helenaford has of course, and I'll pull this locally and run it but I expect it'll run fine in testing

mikehardy commented on tests_react_native/example/app.tsx:
```diff
@@ -237,6 +237,18 @@ function Root(): any {
+          <Button
+            title={`check can create timestamp trigger`}
+            onPress={async (): Promise<void> => {
+              console.log(await Notifee.canCreateTimeStampTrigger());
```
helps keep things straight in the log - noticed while testing it was just true/false, this makes it easier to read

```suggestion
              console.log('Notifee.canCreateTimeStampTrigger: ' + await Notifee.canCreateTimeStampTrigger());
```

mikehardy commented:
worked for me in local testing on ios, android11 and android12 :trophy:

mikehardy commented:
The Android E2E tests are a little flaky which is why I pulled it and ran it locally to double-check. It does work.

The lint check is new, unfortunately my documentation suggestions were okay as far as text goes, but with regard to javascript formatting they need linebreaks and spacing changes to pass the prettier/eslint format rules

Curious for thoughts still on extending the notification status with AndroidNotificationSettings as you suggested

Curious for Helena's thoughts still

Cheers!

vincent-paing commented:
Fixed the lint and added instructions on adding the  `uses-permission` line in `AndroidManifest.xml`

For extending `getNotificationSettings`, I'm still conflicted because this doesn't directly related with notification settings. Alarm is considered as different API in Android. But however, since we are dealing with Alarm API, I do also believe it makes it convenient for the developers. If we agreed to add this, then I'd be happy to submit another PR after this is merged.

mikehardy reviewed (approved):
I agree it's a little bit of mixing of concerns, I'm happy to defer to @helenaford then, and I'm +1 on this, I've tested it locally and it's working for me (so I'm not worried about the android e2e failure at the moment)

helenaford commented on packages/react-native/src/NotifeeApiModule.ts:
```diff
@@ -350,6 +350,20 @@ export default class NotifeeApiModule extends NotifeeNativeModule implements Mod
 
+  public canCreateTimeStampTrigger = (): Promise<boolean> => {
+    if (isAndroid) {
+      return this.native.canCreateTimeStampTrigger();
```
@vincent-paing thank you so much for this contribution. I will be looking at this p/r closer over the next few days. But, one small thing can we make `timestamp` all lowercase... e.g. `canCreateTimestampTrigger`

helenaford replied:
@vincent-paing thank you so much for this contribution. I will be looking at this p/r closer over the next few days. But, one small thing can we make `timestamp` all lowercase... e.g. `canCreateTimestampTrigger`

helenaford commented:
@vincent-paing @mikehardy that sounds great if we can include this in `getNotificationSettings`, I think it makes sense and simplifies the process.

helenaford commented on android/src/main/java/app/notifee/core/AlarmPermissionBroadcastReceiver.java:
```diff
@@ -0,0 +1,19 @@
+
+    if (intent.getAction().equals(ACTION_SCHEDULE_EXACT_ALARM_PERMISSION_STATE_CHANGED)) {
+      Log.i("AlarmPermissionReceiver", "Received alarm permission state changed event");
+      new NotifeeAlarmManager().rescheduleNotifications();
```
Keep this as its own `BroadcastReceiver` sounds better, as I feel it'll give more flexibility in the future

helenaford commented on android/src/main/java/app/notifee/core/Notifee.java:
```diff
@@ -119,6 +120,17 @@ public void cancelAllNotificationsWithIds(
   }
 
+  @KeepForSdk
+  public void canCreateTimeStampTrigger(MethodCallResult<Boolean> result) {
```
@vincent-paing if we extend `getNotificaationSettings`, would we still need this API?

vincent-paing replied:
I think we can keep both and mark this as deprecated to be removed in future in that case? Or do you feel like we should get it right the first time to prevent long tail?

mikehardy replied:
I think these are all new APIs, no need to follow a deprecation cycle, they have never been released so technically did not exist and cannot be deprecated

vincent-paing replied:
@mikehardy For now, we would not need to but would we need to mark this as deprecate if we put this boolean flag in `getNotificationSettings` in future?

mikehardy replied:
Hmm - I guess it comes down to how strongly @helenaford feels about mixing the concerns whether it is necessary to even keep a separate API.

To define terms just a bit since we are all asynchronous and I may not communicate clearly otherwise: by mixing concerns I mean that permission to set alarms is not really permission to send notifications, they are separate technically. 

Now, does that matter, to me?
From a notification module perspective the only use for the alarm is to display a notification, so from that perspective, I think they are close enough to the same thing that I'm okay mixing them. As a separate criteria for whether it's okay or not I think about the developer use case, and I can't imagine a case where I wouldn't want to run all the permissions / capability checks at once, so to me, I think mixing makes sense in this library and for our developers.

So I like the idea of putting the key in notification settings.

At that point, it feels having a separate API does not have much value. So I personally would put the key in notification settings (and have it come back when the API for getting settings is called), and I'd get rid of the separate API

Helena may or may not feel differently :-). And either way, I really want to emphasize that this is in my mind a tiny bit of polish on a PR that's fantastic, and I appreciate the PR a lot, excited to see it merged. Thanks

vincent-paing replied:
I can tell that all of us felt that putting that in notification settings might be more useful for developers than to make it a separate API. In architectural perspective, I will put the native API boolean check into `AlarmUtils` so that it's still logically separated in native side, delete the separate API on the react native part of the lib and merged it into `getNotificationSettings`. This way, we can prevent long tail issue of having to maintain an extra function, and also wouldn't require us to release a breaking change in future when we remove this API completely. 

I might be able to find some free time during weekends and will make changes then

helenaford replied:
thanks @vincent-paing. sounds good. As soon as that's done, I'll hit merge and get it released 😎

vincent-paing commented:
Refactored the check into `getNotificaitionSettings`. While fixing the e2e test, I noticed that it's very specific to the OS version we use. The settings page change on every OS and it will be very flaky. Instead of `UIAutomator`, we might have to look into making cavy either interact with system UI or mock the system api. Or we could have coverage with unit tests on both js and kotlin code. For now, I removed that e2e test since it's very OS specific. Tested manually on both Android and iOS and it's fine for me. 

Let me know if you have any suggestions on variable naming, documentations or the API structure 💯

mikehardy reviewed (approved):
I can't find a single thing, this looks wonderful

helenaford commented on docs/react-native/docs/triggers.md:
````diff
@@ -143,16 +143,43 @@ const trigger: TimestampTrigger = {
+```xml
+<manifest>
+    ...
+    <uses-permission android:name="android.permission.SCHEDULE_EXACT_ALARM"/>
````
@vincent-paing not sure why we need this as we have this already set in the lib's manifest (https://github.com/invertase/notifee/blob/main/android/src/main/AndroidManifest.xml)

vincent-paing replied:
Oh I didn't notice that it was already in the lib. I will remove this from the doc

helenaford replied:
thank you 😎

helenaford commented on docs/react-native/docs/triggers.md:
````diff
@@ -143,16 +143,34 @@ const trigger: TimestampTrigger = {
+
+```js
+const settings = Notifee.getNotificationSettings();
+if (settings.android.allowedTimestampTrigger) {
````
@vincent-paing can we rename this to `alarm` or something more generic? The reason being is that in the future other trigger types e.g. the `IntervalTrigger` could be created with the alarm manager.

vincent-paing replied:
@helenaford I will rename it to `alarm` to be consistent with naming for `iOSSettings`

vincent-paing commented:
I just noticed we're returning `true` or `false` boolean for alarm whereas `iOSSettings` returns `-1`, `0` or `1` since it's enum. Should `alarm` also be either `0` or `1` to make it consistent or should it be boolean since it can never be more than true/false?

mikehardy commented:
google always seems to follow apple about one  or two versions with regard to permissions, I'm they'll offer a provisional one later too ;-). I favor consistency and same enum if possible, personally. What do you think?

vincent-paing commented:
@mikehardy I think you might be talking about notification permission in general. In that case yes I think there will be provisional later. For `alarm` however, I don't think there will be provisional one. And since alarm has been around since API level 1, there won't be `NOT_SUPPORTED` enum either, all Android versions supports creation of alarm.

mikehardy commented:
Oh yes - I was conflating the two, apologies. Still seems best to go for consistency. Would you believe Android TV doesn't support notifications? It doesn't unless you are a system app, if I understand correctly (though it just quietly drops them with a logcat message, again, if I recall correctly), so NOT_SUPPORTED is actually kind of valid for the android platform though we don't handle TV use cases well at all at the moment

mikehardy commented:
+1 for consistency, I like it

helenaford commented:
This looks great! merged Will release shortly. @vincent-paing thank you for this contribution 😎 🔥 🚀
